tests/NuGetGallery.Facts/Services/PackageMetadataValidationServiceFacts.cs

```diff
@@ -1355,6 +1355,211 @@ private async Task<PackageValidationResult> ValidatePackageWithIcon(string iconP
+            }
+
+            private async Task<PackageValidationResult> ValidatePackageWithReadme(string readmePath, byte[] readmeFileData)
+            {
+                _nuGetPackage = GeneratePackageWithUserContent(
+                    readmeFilename: readmePath,
+                    readmeFileBinaryContents: readmeFileData,
+                    licenseExpression: "MIT",
+                    licenseUrl: new Uri("https://licenses.nuget.org/MIT"));
+                _featureFlagService
+                    .Setup(ffs => ffs.IsUploadEmbeddedReadmeEnabled(_currentUser))
+                    .Returns(true);
+
+                return await _target.ValidateMetadataBeforeUploadAsync(
+                    _nuGetPackage.Object,
+                    GetPackageMetadata(_nuGetPackage),
+                    _currentUser);
+            }
+
+            [Theory]
+            [InlineData("", false)]
+            [InlineData(".txt", false)]
+            [InlineData(".md", true)]
+            [InlineData(".Txt", false)]
+            [InlineData(".Md", true)]
+            [InlineData(".doc", false)]
+            [InlineData(".pdf", false)]
+            public async Task ChecksReadmeFileExtension(string extension, bool successExpected)
+            {
+                string readmeFileName = $"readme{extension}";
+
+                var result = await ValidatePackageWithReadme(readmeFileName, new byte[] { 0xFF, 0xD8, 0xFF, 0x32 });
+
+                if (successExpected)
+                {
+                    Assert.Equal(PackageValidationResultType.Accepted, result.Type);
+                    Assert.Null(result.Message);
+                }
+                else
+                {
+                    Assert.Equal(PackageValidationResultType.Invalid, result.Type);
+                    Assert.Contains(extension, result.Message.PlainTextMessage);
+                    Assert.Contains("invalid extension", result.Message.PlainTextMessage);
+                    Assert.Contains("md", result.Message.PlainTextMessage);
+                }
+                Assert.Empty(result.Warnings);
+            }
+
+            [Fact]
+            public async Task RejectsLongReadme()
+            {
+                const int ExpectedMaxReadmeLength = 1024 * 1024;
+
+                var readmeTextBuilder = new StringBuilder(ExpectedMaxReadmeLength + 100);
+
+                while (readmeTextBuilder.Length < ExpectedMaxReadmeLength + 1)
+                {
+                    readmeTextBuilder.AppendLine("Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.");
+                }
```

Comment:
Would something like this work instead?

```suggestion
                var readmeText = new string('a', ExpectedMaxReadmeLength + 100);
```

See: https://docs.microsoft.com/en-us/dotnet/api/system.string.-ctor?view=netcore-3.1#System_String__ctor_System_Char_System_Int32_

Reply:
I think it will work